src/twisted/internet/defer.py

```diff
@@ -195,6 +231,8 @@ def maybeDeferred(
         return result
     elif isinstance(result, Failure):
         return fail(result)
+    elif _isAsyncDefValue(result):
+        return Deferred.fromCoroutine(result)
```

Comment:
Pushing this into a separate function seems oddly asymmetrical here, and given that we're already in for a penny here on `isinstance` calls, it seems like we might as well be in for a pound?  We already dropped the `type(result) is Deferred` optimization many years ago because it raised too many odd edge cases.

I'd rather leave this as a standard-looking `isinstance` ladder to make it more obvious that some optimization (maybe `singledispatch` will be super fast one day?) could be applied, and remove the indirection.

```suggestion
    elif isinstance(result, CoroutineType):
        return Deferred.fromCoroutine(result)
```

Reply:
(Also this raises an interesting question: what about `Future`s 🤔)

Reply:
I did have the check inline originally but that left me without a very satisfying place to put the large comment.  If I move the comment back I guess it breaks up the aesthetic of that nice wish-it-were-a-pattern-match structure. :)

Also I guess I originally imagined the function might have another caller in this module but that seems less likely now that I've looked at more of the pieces involved.  Happy to move it back, if you like.

I'm not sure about `isinstance(result, CoroutineType)` instead of `type(result) is CoroutineType` though.  If the idea is to support exactly whatever `async def ...` gives you then I think the latter is more correct.  The latter means that `__instancecheck__` and `__subclasscheck__` are potentially getting involved, right?

Speaking of that, that `Future` question is really a can of worms.  If `CoroutineType`, why not `Future`?  If `Future`, why not `Awaitable`?

I think the answer is that the goal of this change is to provide a transition path from "Deferred-or-not" to "async def" for certain interfaces.  The goal is _not_ "make maybeDeferred compatible with all of the random kinds of async operation APIs" - because maybeDeferred itself is a bit of a misfeature, we're not trying to make it great and powerful, just to provide a path to completely escape from it - and supporting "async def" is enough for that.

Something we should discuss making great and powerful, though, might be `Deferred._runCallbacks` ... should a Deferred callback be allowed to return any Awaitable?  But that is distinctly a separate question.

What do you think?

Reply:
I'd say stick with `type(result) is CoroutineType` as it's easy to relax that to an `isinstance()` check if necessary, but difficult to optimize in the other direction.

I'd also mildly prefer it were inline, but I think it's good either way.

Reply:
now inline, w/ still using `type(result) ...`